### DISCUSSION TITLE
I'm considering adding TiDB support ...

### DISCUSSION THREAD

**Discussion by dankasak** [photoprism/photoprism#3218] [General]
Hi there. I'm Dan, a software/data engineer / DBA, currently working as a support engineer / DBA / TAM for PingCAP, the authors of TiDB. As I already run a bunch of TiDB clusters ( including on my home server, where I'm setting up photoprism ), it makes sense for me to put everything MySQL-compatible into it, and generally play around with it, trying to break things.

As such, I've attempted to bring up Photoprism on TiDB, and hit some small issues ( eg ticket #3199 and ticket #3216 ... which I submitted a patch for in #3217 ). I see there are a few more issues, like "create or replace" syntax for indexes that don't parse on TiDB ( the "or replace" part isn't parsed ), and things that could be worked around by setting a SQL_MODE session variable upon connecting ( eg no_zero_in_date ).

I assume there would only be a small number of further fixes until it's working. I'm interested in completing this, and wondering:

- Would there by any opposition to me merging such work? I know there is a "preference" for people to not use MySQL, and I'm not sure how strong that is, and if it extends to other databases. I would of course commit to maintaining such support, and would be somewhat motivated to do so, since it would be hosting all of my personal photos.
- The easiest way to approach this ( from my perspective ) would be to tweak existing "MySQL" bits to be more MySQL/TiDB friendly. But I assume this would **not** be the preferred approach from others. The alternative is adding TiDB as a separate database type, which would be the least risk for existing MariaDB support stability. Am I correct in this assumption? :) If so ... I'm not particularly familiar with Go, and would appreciate a high-level summary of what would be required to complete this.

**Comment by lastzero** [comment]
It seems feasible to add TiDB as a separate dialect and use a switch where needed. 

Note that we have supported TiDB in the past:

https://github.com/photoprism/photoprism/blob/240393d8b60da002e4486559263861545fe354fd/internal/config/config_db.go#L46

There just wasn't enough time to maintain it, and there hasn't been much demand for it either. So if we were to reintroduce TiDB support, this would require ongoing help with maintenance and testing.

**Reply by dankasak** [reply]
Got it, thanks. I'm still interested in working on this. Would you be opposed to **not** including TiDB, and instead requiring users who wanted to use TiDB maintain their own cluster? Rationale:

- TiDB is a distributed database, and should really be deployed separately, on multiple nodes - this is the intended architecture.
- Regular users are not going to be particularly interested in running TiDB over MariaDB. TiDB would mainly benefit large deployments ( eg multi-tenanted ) or users needing fault tolerance / healing ( eg can survive individual nodes going down ).
- TiDB has a 'tiup' utility, which is a package and cluster manager. This is the easiest way of bringing up a cluster - including for unit tests, or a production cluster.
- Keeping TiDB components separate from Photoprism keeps Go library management much simpler.
